`generate.py` changes

diivm opened the pull request:
- Part of splitting #8 (3/N; N=3)

diivm committed:
[generate_bindings.py] initial changes; building

some more progress

still working

some more

a little more

some more

it generates!!

changes

diivm committed:
[generate_bindings.py] stuff, not working as of yet

diivm committed:
[generate_bindings.py] fixed some stuff, needs cleaning

diivm committed:
[generate_bindings.py] cmake build without error!

diivm committed:
[generate_bindings.py] clean

diivm committed:
[generate.py] use utils

diivm committed:
[generate.py] messy fix for anonymous gen and member definition

diivm committed:
[generate.py] add inclusion_directive

diivm committed:
[generate.py] fix cxx_method

diivm committed:
[generate.py] remove parameter_list from constructor

diivm committed:
[generate.py] some comments and minor fixes

diivm committed:
[test_generate.py] init file

diivm committed:
[generate.py] comment updation and minor refactoring

diivm committed:
[generate.py] list to tuple `in`s

diivm committed:
[generate.py] new names

diivm committed:
[generate.py] remove `get_prev_depth_node`

diivm committed:
[generate.py]  add docstrings - 1

diivm committed:
[generate.py] add libraries everytime

diivm committed:
[generate.py] unnecessary if

diivm committed:
[generate.py] close() -> end_scope() rename

diivm committed:
[generate.py] make `get_fields_from_anonymous` static

diivm committed:
[generate.py] `get_fields_from_anonymous` better internals

diivm committed:
[generate.py] extract generation for testing calls

diivm committed:
[test_generate.py] intial setup and some tests

diivm committed:
[generate.py] elif fix

diivm committed:
[generate.py] separate pybind line list

diivm committed:
[generate.py] fix hardcoded filepath

diivm committed:
[test_generate.py] fix intial_lines problem, new function

diivm committed:
[generate.py] add template support

bugs encountered:
- extra constructor at the beginning of file pcl_base.cpp gives rise to standalone .def
- inclusion of cpp files: header files must be included, not the cpp file
- namespace pcl should be in all code
- `;` at the end of class not generated if no defs

diivm committed:
[generate.py] minor fix, use class_name

diivm committed:
[test_generate.py] updates

- rename `get_character_only_string` -> `remove_whitespace`
- functional changes to `remove_whitespace`
- rename `get_binded_code` -> `generate_bindings`
- tmp_path change order and comment updation
- dump_json: separators for json size reduction
- include: remove pcl
- pybind_linelist -> initial_pybind_lines
- default constructor

diivm committed:
[utils.py] add options to dump_json

diivm committed:
[generate.py] renaming and change `pcl/filename`->`filename` in include

diivm committed:
[test_generate.py] str won't work

diivm committed:
[generate.py] if condition for default ctor in handle_constructor

aPonza commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,352 @@
+    pybind_linelist = [
+        "#include <pybind11/pybind11.h>",
+        "#include<pybind11/stl.h>",
+        "#include<pybind11/stl_bind.h>",
```
add space between `#include` and the header in both these

aPonza commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,352 @@
+            "CXX_TRY_STMT": handled_by_pybind,
+            "CXX_CATCH_STMT": handled_by_pybind,
+            "CLASS_DECL": self.handle_struct_decl,
+        }
```
Can you reorder these? Maybe private (leading `_`), public, vars, and within `kind_functions` in some meaningful way? IIRC kunal was already mentioning this in the old aggregate PR.

diivm replied:
Did. Let me know if alternate ordering needed.
See d31691f62d62b1ce2f21edf96a819038b05237da (at least before rebase :joy: )

aPonza commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,352 @@
+                )
+                flag = True
+        if not flag:
+            self.linelist.append(f"template<>")
```
Does using a generator expression like
```
if not any(sub_item["kind"] == "TEMPLATE_NON_TYPE_PARAMETER" for sub_item in self.members):
  self.linelist.append(f"template<>")
else:
  <you'd still need to loop around sub_items for their element type and name here>
```
help in the way of removing `flag`? I don't particularly love it.

diivm replied:
This function is useless and will be removed in another PR (I am trying to minimize changes in this PR as this is already a pretty long one).

diivm replied:
Removed.

aPonza commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,352 @@
+        # for inclusion in self.inclusion_list:
+        #     final.append(f"#include <{inclusion}>")
+        final += self.pybind_linelist
+        for i in range(len(self.linelist)):
+            if self.linelist[i].startswith("namespace"):
+                continue
+            else:
+                self.linelist[i] = "".join(
+                    (f"PYBIND11_MODULE({module_name}, m)", "{", self.linelist[i])
+                )
+                break
```
```suggestion
        for line in self.linelist:
            if line.startswith("namespace"):
                continue
            else:
                line = "".join(
                    (f"PYBIND11_MODULE({module_name}, m)", "{", line)
                )
                break
```

diivm replied:
I'll address this (along with other function's internal's changes) in another PR.

diivm replied:
@aponza
We want to change `self._linelist[i]`, not its `line` (strings are immutable -> copy created)

We need index based:
```py
self.linelist[i] = "".join((f"PYBIND11_MODULE({module_name}, m)", "{", self.linelist[i]))
```

aPonza replied:
not sure I got you, still, if you need an index, don't use `for i in range(len(x))` but use `for i, _ in enumerate(x)`

diivm replied:
Oh yeah, that was really vague what I wrote :joy: 
Better explained via code:
```py
>>> str_list = ["str0", "str1", "str2"]
>>> compare_list = []



>>> for item in str_list:
...     if item == "str1":
...         item = "str100"
...     compare_list.append(item)
... 
>>> print(compare_list)
['str0', 'str100', 'str2']
>>> print(str_list)
['str0', 'str1', 'str2'] <------------------------------------------ not updated



>>> for i in range(len(str_list)):
...     if str_list[i] == "str1":
...         str_list[i] = "str100"
... 
>>> print(str_list)
['str0', 'str100', 'str2']
```

I'll do `enumerate` :+1:

aPonza replied:
Now I get what you meant and of course it made sense! Cheers

diivm replied:
Done.

diivm committed:
[generate.py] documentation and rearranging

diivm committed:
Merge pull request #40 from divmadan/fix/make_lives_easier-2

`generate.py` tests

diivm committed:
[generate.py] fix bad merge and an elif chain

divmadan commented:
Commits
```
95016c29a Merge pull request #40 from divmadan/fix/make_lives_easier-2
9a70fa1a6 [generate.py] if condition for default ctor in handle_constructor
ba275b729 [test_generate.py] str won't work
86b10f019 [generate.py] renaming and change `pcl/filename`->`filename` in include
5875c31fd [utils.py] add options to dump_json
28a16733f [test_generate.py] updates
12617c23e [test_generate.py] fix intial_lines problem, new function
232dd9844 [test_generate.py] intial setup and some tests
0d09658b8 [test_generate.py] init file

```
doesn't need a review as they came via #40 merge

diivm committed:
[generate.py] comments and make pybind lines private

diivm committed:
[generate.py] updation

- comments
- rename cxx_base_specifier_list -> base_class_list
- simplify working of base_class_list
- elif fixes

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -111,9 +118,11 @@ def end_scope(self):
             self._linelist.append(";")
-        if self._state_stack[-1]["kind"] == "CLASS_TEMPLATE":
+        elif self._state_stack[-1]["kind"] == "CLASS_TEMPLATE":
             self._linelist.append(";")
```
`STRUCT_DECL` is repeated twice. 

```py
def end_scope(self):
    kind = self._state_stack[-1]["kind"]
	end_token = collections.default_dict(str)
    # if you use a normal dict, then you need to use `end_token.get(kind, '')`

	end_token['NAMESPACE'] = '}'
	end_token['CLASS_TEMPLATE'] = ';'
	end_token['STRUCT_DECL'] = ';'

	self._linelist.append(seperator[kind])
```

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -5,77 +5,83 @@
-            "CXX_NULL_PTR_LITERAL_EXPR": no_need_to_handle,
-            "STRING_LITERAL": no_need_to_handle,
+            "VARIABLE_REF": handled_by_pybind,
+            # STMTs
```
Don't skimp on letter count in comments

diivm commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,417 @@
+        # TODO: Merge this and next block via a design updation
+        # handle anonymous structs, etc. as field declarations
+        for sub_item in self.members:
+            fields = self.get_fields_from_anonymous(sub_item)
+            for field in fields:
+                if field["element_type"] == "ConstantArray":
+                    # TODO: FIX: readwrite, not readonly
+                    self._linelist.append(
+                        f'.def_property_readonly("{field["name"]}", []({self.name}& obj) {{return obj.{field["name"]}; }})'  # float[ ' + f'obj.{sub_item["name"]}' + '.size()];} )'
+                    )
+                else:
+                    self._linelist.append(
+                        f'.def_readwrite("{field["name"]}", &{self.name}::{field["name"]})'
+                    )
```
I get a feeling that this could be 2 separate functions:
* handle_anon
* handle_fields

and you'd just call them like:
```python
for elem in self.members:
    self.kind_functions[elem["kind"](elem)
```

_Originally posted by @kunaltyagi in https://github.com/divmadan/pcl/pull/8_

diivm replied:
Well, the initial plan was that each kind is handled like we handle class, structs, etc. That's how I was developing until a point where it became error-prone (parent access required, etc.).
Via these lines:
https://github.com/divmadan/pcl/blob/d7f8f4b67d8dff6c4981698927bdd1bcdc1852ac/bindings/python/scripts/generate.py#L177-L183

I'll try to separate in a more abstract manner (see how fields are binded and try to club them, maybe a static method)

diivm commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,417 @@
 
-def handle_alias(item):
-    pass
+            # handle field declarations
+            if sub_item["kind"] == "FIELD_DECL":
+                if sub_item["element_type"] == "ConstantArray":
```
Could you give an example of constant array?

_Originally posted by @kunaltyagi in https://github.com/divmadan/pcl/pull/8_

diivm replied:
A C-style array.
```cpp
#define PCL_ADD_UNION_POINT4D \
  union EIGEN_ALIGN16 { \
    float data[4]; \   <------------------------------------------- THIS
    struct { \
      float x; \
      float y; \
      float z; \
    }; \
  };
```

```json
              {
                  "depth": 4,
                  "line": 291,
                  "column": 5,
                  "kind": "FIELD_DECL",
                  "name": "data",
                  "element_type": "ConstantArray",
                  "access_specifier": "PUBLIC",
                  "members": [
                    {
                      "depth": 5,
                      "line": 291,
                      "column": 5,
                      "kind": "INTEGER_LITERAL",
                      "name": "",
                      "element_type": "Int",
                      "members": []
                    }
                  ]
                }
```

diivm replied:
It's read only (accessing it as an object) as of now, have to figure out how to write to it.

kunaltyagi replied:
Mark as todo somewhere

diivm replied:
Already marked.

diivm commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,417 @@
-        type_functions[sub_item["type"]](sub_item)
-    module_linelist.append(";")
-    in_struct = False
+        # TODO: update blacklist
+        # blacklist = ["pcl/memory.h", "pcl/pcl_macros.h",]
```
Why a blacklist?

_Originally posted by @kunaltyagi in https://github.com/divmadan/pcl/pull/8_

diivm replied:
We don't need one? To exclude binding of `macros.h` and alike?

kunaltyagi replied:
Macros will have no code to bind, right?

(pcl_config.h might have macro constants that we want to bind)

diivm replied:
If that's the case, then no need. I was under the assumption that we may have some files that we want to skip. 

Removed.

diivm commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,417 @@
+        Handle `CursorKind.INCLUSION_DIRECTIVE`
+        """
+
+        # TODO: develop
+        pass
 
-    module_linelist.append(f'py::class_<{name_impl}>(m, "{name_decl}")')
-    for sub_item in item["members"]:
-        type_functions[sub_item["type"]](sub_item)
-    module_linelist.append(";")
-    in_struct = False
+        # TODO: update blacklist
+        # blacklist = ["pcl/memory.h", "pcl/pcl_macros.h",]
+        # if self.name.startswith("pcl"):
+        #     self._inclusion_list.append(self.name)
```
Is it possible to factor out `"pcl"` to the class ctor?

_Originally posted by @kunaltyagi in https://github.com/divmadan/pcl/pull/8_

diivm replied:
Done.

diivm commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,417 @@
+                fields.append(sub_item)
+            # recurse
+            elif sub_item["kind"] in ("ANONYMOUS_UNION_DECL", "ANONYMOUS_STRUCT_DECL"):
+                fields += bind.get_fields_from_anonymous(sub_item)
```
why change from self to bind?
```py
class A:
    def __init__(self):
        self.say_hi()
    @staticmethod
    def say_hi():
        print("I'm a static method, short and stout")

A()
A.say_hi()
```

_Originally posted by @kunaltyagi in https://github.com/divmadan/pcl/pull/8_

diivm replied:
It's being called from a class function:
https://github.com/divmadan/pcl/blob/d7f8f4b67d8dff6c4981698927bdd1bcdc1852ac/bindings/python/scripts/generate.py#L128-L148

Why create an object of class within the function?

kunaltyagi replied:
`self` should work. If it doesn't that's ok (resolve this convo). Else using `bind` makes renaming the class in future difficult

diivm replied:
Stupid pylint!
It works without any `self` or classname :facepalm: 

Updated.

diivm reviewed (commented):
Remaining review comments from #8

diivm committed:
[test_generate.py] initial_pybind_list private

diivm committed:
[generate.py] remove handle_class_template

diivm committed:
[generate.py] better end_scope

diivm committed:
[generate.py] expand comments

diivm committed:
[generate.py] module_name to class ctor

diivm committed:
[test_generate.py] module_name to class ctor

diivm committed:
[generate.py] unnecessary collections import

diivm committed:
[generate.py] remove `handle_final` from class

- moved it to the generate function, as a nested function `combine_lines`

diivm committed:
[generate.py] move combine_lines out of the if loop

diivm committed:
[generate.py] Add option of directly operating on dicts apart from JSONs

diivm committed:
[test_generate.py] `generate_bindings`: operate directly on dict

diivm committed:
[generate.py] minor changes

- add type bounds
- comments
- named args

kunaltyagi reviewed (commented):
No major points except the previously raised ones.

diivm committed:
[generate.py] range to enumerate

diivm committed:
[generate.py] remove self from static method

diivm committed:
[generate.py] remove blacklist

diivm committed:
[generate.py] add Exception to return type

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,104 @@
+    }
+    """
+
+    assert output == get_expected_string(expected_module_code=expected_module_code)
```
Big areas of missing tests: (add as todo somewhere):
* function
* templated and instantiated function
* templated and instantiated class
* anonymous struct
* enum

Anything else?

diivm replied:
Created an issue

kunaltyagi replied:
Did you note this somewhere?

aPonza commented on bindings/python/scripts/generate.py:
```diff
@@ -1,105 +1,404 @@
+        parameter_type_list = []
+
+        # generate parameter type list
+        for sub_item in self.members:
+            if sub_item["kind"] == "PARM_DECL":
+                if sub_item["element_type"] == "LValueReference":
+                    for sub_sub_item in sub_item["members"]:
+                        if sub_sub_item["kind"] == "TYPE_REF":
+                            # @TODO: Make more robust
+                            type_ref = (
+                                sub_sub_item["name"]
+                                .replace("struct ", "")
+                                .replace("pcl::", "")
+                            )
+                            parameter_type_list.append(f"{type_ref} &")
+                elif sub_item["element_type"] == "Elaborated":
+                    namespace_ref = ""
+                    for sub_sub_item in sub_item["members"]:
+                        if sub_sub_item["kind"] == "NAMESPACE_REF":
+                            namespace_ref += f'{sub_sub_item["name"]}::'
+                        if sub_sub_item["kind"] == "TYPE_REF":
+                            parameter_type_list.append(
+                                f'{namespace_ref}{sub_sub_item["name"]}'
+                            )
+                elif sub_item["element_type"] in ("Float", "Int"):
+                    parameter_type_list.append(f'{sub_item["element_type"].lower()}')
+                else:
+                    parameter_type_list.append(f'{sub_item["element_type"]}')
+        parameter_type_list = ",".join(parameter_type_list)
```
This is begging for a refactor, too much nesting. Make a note to extract some `handle_` functions.

aPonza commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,104 @@
+    return expected_output
+
+
+def test_case1(tmp_path):
```
```suggestion
def test_struct_without_members(tmp_path):
```

aPonza commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,104 @@
+    assert output == get_expected_string(expected_module_code=expected_module_code)
+
+
+def test_case2(tmp_path):
```
```suggestion
def test_struct_with_members(tmp_path):
```

diivm committed:
[generate.py] nothing, self doesn't work; use classname in staticmethod

diivm committed:
[generate.py] comments

diivm committed:
[test_generate.py] rename
